src/v2/catalog.rs

```diff
@@ -1,60 +1,61 @@
+use async_stream::try_stream;
+use futures::stream::Stream;
+use futures::{self};
+use reqwest::{RequestBuilder, StatusCode};
 
 #[derive(Debug, Default, Deserialize, Serialize)]
 struct Catalog {
     pub repositories: Vec<String>,
 }
 
 impl v2::Client {
-    pub fn get_catalog(&self, paginate: Option<u32>) -> StreamCatalog {
-        let url = {
-            let suffix = if let Some(n) = paginate {
-                format!("?n={}", n)
-            } else {
-                "".to_string()
+    pub fn get_catalog<'a, 'b: 'a>(
+        &'b self,
+        paginate: Option<u32>,
+    ) -> impl Stream<Item = Result<String>> + 'a {
+        try_stream! {
+            let url = {
+                let suffix = if let Some(n) = paginate {
+                    format!("?n={}", n)
+                } else {
+                    "".to_string()
+                };
+                let ep = format!("{}/v2/_catalog{}", self.base_url.clone(), suffix);
+
+                reqwest::Url::parse(&ep).map_err(|e|
+                        Error::from(format!(
+                            "failed to parse url from string '{}': {}",
+                            ep, e
+                        ))
+                )?
             };
-            let ep = format!("{}/v2/_catalog{}", self.base_url.clone(), suffix);
-            match reqwest::Url::parse(&ep) {
-                Ok(url) => url,
-                Err(e) => {
-                    return Box::new(stream::once::<_, _>(Err(Error::from(format!(
-                        "failed to parse url from string '{}': {}",
-                        ep, e
-                    )))));
-                }
-            }
-        };
+            let req = self.build_reqwest(reqwest::Client::new().get(url));
+
+            let catalog = fetch_catalog(req).await?;
 
-        let req = self.build_reqwest(reqwest::r#async::Client::new().get(url));
+            for repo in catalog.repositories {
+                yield repo;
+            }
+        }
+    }
+}
 
-        let fres = req
-            .send()
-            .from_err()
-            .and_then(|r| {
-                let status = r.status();
-                trace!("Got status: {:?}", status);
-                match status {
-                    StatusCode::OK => Ok(r),
-                    _ => Err(format!("get_catalog: wrong HTTP status '{}'", status).into()),
-                }
-            })
-            .and_then(|r| {
-                r.into_body().concat2().map_err(|e| {
-                    format!("get_catalog: failed to fetch the whole body: {}", e).into()
-                })
-            })
-            .and_then(|body| -> Result<Catalog> {
-                serde_json::from_slice(&body).map_err(|e| e.into())
-            })
-            .map(|cat| futures::stream::iter_ok(cat.repositories.into_iter()))
-            .flatten_stream();
-        Box::new(fres)
+async fn fetch_catalog(req: RequestBuilder) -> Result<Catalog> {
+    match req.send().await {
```

Comment:
I think a few indent-levels in this block logic can be dropped by immediately `?`ing the intermediate results. That should make the happy-path logic more obvious to follow.